src/jupyterlabcontroller/services/lab.py

```diff
@@ -0,0 +1,204 @@
+
+        #
+        # This process has three stages: first is the creation or recreation
+        # of the user namespace.  Second is all the resources the user Lab
+        # pod will need, and the third is the pod itself.
+        #
+
+        await self.create_user_namespace()
+        await self.create_user_lab_objects()
+        await self.create_user_pod()
+
+    async def create_user_namespace(self) -> None:
+        ns_retries: int = 0
+        ns_max_retries: int = 5
+        ns_name = self.context.namespace
+        try:
+            await self.nublado.k8s_client.create_namespace(ns_name)
+        except NSCreationError as e:
+            if e.status == 409:
+                self.nublado.logger.info(f"Namespace {ns_name} already exists")
+                # ... but we know that we don't have a lab for the user,
+                # because we got this far.  So there's a stranded namespace,
+                # and we should delete it and recreate it.
+                #
+                # The spec actually calls for us to delete the lab and then the
+                # namespace, but let's just remove the namespace, which should
+                # also clean up all its contents.
+                await self.nublado.k8s_client.delete_namespace(ns_name)
+                ns_retries += 1
+                # Give up after a while.
+                if ns_retries > ns_max_retries:
+                    raise RuntimeError(
+                        "Maximum namespace creation retries "
+                        f"({ns_max_retries}) exceeded"
+                    )
+                # Just try again, and return *that* one's return value.
+                return await self.create_user_namespace()
+            else:
+                self.nublado.logger.exception(
+                    f"Failed to create namespace {ns_name}: {e}"
+                )
+                raise
+
+    async def create_user_lab_objects(self) -> None:
+        # Initially this will create all the resources in parallel.  If it
+        # turns out we need to sequence that, we do this more manually with
+        # explicit awaits.
+        scheduler: Scheduler = Scheduler(
+            close_timeout=self.nublado.config.kubernetes.request_timeout
+        )
+        await scheduler.spawn(self.create_secrets())
+        await scheduler.spawn(self.create_nss())
+        await scheduler.spawn(self.create_env())
+        await scheduler.spawn(self.create_network_policy())
+        await scheduler.spawn(self.create_quota())
+        await scheduler.close()
+        return
+
+    async def create_secrets(self) -> None:
+        data: Dict[str, str] = await self.merge_controller_secrets()
+        await self.nublado.k8s_client.create_secret(
+            name=f"nb-{self.context.user.username}",
+            namespace=self.context.namespace,
+            data=data,
+        )
+
+    async def merge_controller_secrets(self) -> Dict[str, str]:
```

Comment:
I would characterize extraction of specific values a business logic decision, so I think I don't have a good model of where your storage layer ends and the service layer begins.

Reply:
Business logic means that the code encapsulates some sort of decision about how something should work specific to the problem that one is trying to solve. For example, the decision that a lab will have two config maps, a secret, a pull secret, etc. is business logic. The decision that all the contents of the secret will come from other secrets in the lab-controller namespace (as opposed to configuration or environment variables or whatever) is business logic.

"Read this list of Kubernetes secrets, extract the values of the given keys, and return them as a dict" is not business logic because there are no decisions in that. It's a Kubernetes API operation that you can mentally mechanically translate into a set of API calls. So that operation would go into the storage layer, whose purpose is to hide the Kubernetes API from the rest of the program.

This distinction is a little subtle, but the idea is that the storage layer contains mechanical operations on some underlying API, and the API between the service and the storage layer is designed to keep the decisions in the service layer but the mechanics in the storage layer so that you can replace the mechanics. So if, for instance, we decided that we wanted to support getting secrets directly from Vault, the service layer would decide whether to ask Vault for secrets rather than Kubernetes, but the code to actually get the secrets out of Vault would be in a Vault storage object and the code to get the secrets out of Kubernetes would be in a Kubernetes storage layer.

Reply:
Done (should do PodSpec too)